### DISCUSSION TITLE
Modulate heating power

### DISCUSSION THREAD

**Discussion by svanscho** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#261] [Ideas]
Hi all, 

What a great project! Keep up the good work, I was so stoked when I found out about this project. 
Exactly what I needed!

After a lot of playing around (I have a background as electronics engineer and work for a Domotica company) I have one more feature I would like to build: modulate the heating power. Where I live (in Belgium, but I know there are many other countries) we get feed-in tarrifs for our solar energy which are much lower than the consumption tariffs. Therefore I have modulated my SPA to mainly heat on solar energy whenever possible. There is only 1 problem: the maximum output of my solar inverter is 1500W. The heating element of the SPA is 2000W. So even in an ideal world where I only heat the SPA when the sun is shining, I still have a 500W base consumption.

I could build some electronics to modulate the heating (e.g. phase cut dimmer) but would like to avoid fiddling with the electronics close to the water as much as possible.

@visualapproach Hence the following question: is there any API/method you know of where we can leverage the CIO to modulate this power? If I can bring it down to around 1000-1500W this would greatly improve my power bill :-)

Thanks in advance for all the help!

**Comment by ChadH360** [comment]
I noticed today, that on the back of my pump unit, it states that heating consumes 2050W, and heating+bubbles consumes 1850W - seems odd that heating+bubbles is lower power than heating only.
My guess is that the heater is actually made up of 2 x 1kw heaters. When heating only, it runs both heaters (+50W for the filter pump).
And when bubbles come on, it just runs 1 heater and the air pump.

Pure speculation, but it might help with a solution to what you're trying to do.

PS I got my WiFi module in my 2021 Maldives today. Sooooo chuffed! It's absolutely awesome.
Don't forget to send visualapproach a coffee or 10 if this awesome project has benefited you :)

**Comment by visualapproach** [comment]
Hi @svanscho ! Great question. It depends on what pump you have. Mostly the 6-wire systems is considered better because we can know what the display shows, and take actions. But in this case you're lucky if you have a 4-wire, because then we can control both heater elements individually. The 6-wire versions is just ON or OFF and when we simulate a press on HEATER ON it does it's thing without any means to control it. Some speculation here: There is two heaters as @ChadH360 said. I haven't examined it too much, but there might be some ways to exploit it's behaviour. Like disconnecting the airbubbles pump and always start airbubbles when heating. Possibly the CIO will think it's on and only start one heater element. The other thing would be to confirm that heater 1 starts some fixed amount of time ahead of heater 2, to avoid power surges. In that case maybe one could turn on heat for that amount of time, then turn it off briefly and turn it on again, and so on.

@ChadH360 Thank you for your support!

**Reply by tunnus-1** [reply]
Hi @visualapproach, as a 4-wire model can control both heater elements individually, was it on your todo list that there would be support for it in the UI? Even MQTT support or similar would do, as I would mainly control spa functions via HA. That would be great as it would allow better optimization for solar power. 

Whether or not that is ever implemented, I'll will definitely buy you a couple of coffees when I've put this wifi control into use (still waiting for pcb & other parts to arrive)

**Reply by visualapproach** [reply]
@tunnus-1 to be honest it's not on my actual todo list, but I've had it in the back of my head all the time. Maybe one day it will happen, and since there is now a request for it, that day is somewhat closer. Have a nice day!

**Reply by visualapproach** [reply]
Now I know why it wasn't on the todo list. It was already done 😁 Although with a bug. See #529

**Reply by GSzabados** [reply]
@visualapproach, does this work for the 6 wire Maldives as well? If yes, then I have to do on my todo list to update the firmware of the ESP to use it. 😃

**Reply by visualapproach** [reply]
No only for 4-wires unfortunately.

**Reply by GSzabados** [reply]
Thanks! I would love to hear a yes, but nothing can be perfect.

**Comment by svanscho** [comment]
Thanks guys. I'll explore the double heaters.

I can already confirm my Helsinki SPA does not do any power surge protection. The graph below is not that clean as the P1 is my total consumption of the house, but you get the idea:
- The blue line is the solar power production
- The green line is the net production. You can clearly see that from the moment I turn on the heaters, it jumps to around 2kW instantly and stays like that all the time.

What I did:
- I turn on the SPA at 34C around 9AM
- SPA turns off around 1PM
- (intermittent glitch of dishwasher)
- I set the temp to 37C around 4:50PM


<img width="1741" alt="image" src="https://user-images.githubusercontent.com/7288358/167203927-c77172ac-4d9b-4b0b-bd39-f63a87e8a42c.png">

**Reply by visualapproach** [reply]
I forgot the obvious solution to disconnect one of the heaters.

**Reply by admiralroflknife** [reply]
I'm happy I procrastinated on installing this till now. I wanted to run my spa off solar as well but was running into the limit of the inverter as well. I'm happy to report that I disconnected one of the heaters on my unit and the power draw dropped to 650w when heating and running the pump (6 wire model). Now with regards to svanscho's issue. I haven't messed with the accessory stuff much but couldn't you theoretically just disconnect one heater and apply a relay between it and the heater and then control it separately? I haven't messed with the accessory stuff much in this project but I think running a relay would be well within its scope. Or you could just do it separately through something like tasmota or home assistant. 

visusalapproach, thanks again for doing all this. It tied into home assistant perfectly through mqtt and in 15 minutes my tub is smart. Very happy!

**Reply by svanscho** [reply]
@admiralroflknife interesting! Do you have any pictures of the heating element you disconnected? 650W seems like a strange number however. I'd expected more something slightly above 1k-ish. Does this mean the heating elements are not symmetric/equal?

On a slightly different note I am currently still investigating how I could truly modulate the heating elements. A simple relay allows for on-off modulation but I'd like to go for the full spectrum using a triac or some other IC of some sorts. To be continued...

**Reply by admiralroflknife** [reply]
I didn't take any pictures unfortunately but this one shows my unit (which internally is very similar to older ones) https://user-images.githubusercontent.com/13341477/152623167-b338481e-d548-47f5-b02f-8ee8466fa82a.png

I unscrewed the 2 screws holding the black cover on right which reveals screw terminals with spade connectors. I disconnected both of the spades for the upper heater. (which are the 4 thin white wires going to the 2 metal bar on the front left of the picture). 

Interesting to go a more complex route. If you could rig up some sort of PID control scheme instead of the current default bang-bang approach that would most likely allow for a much reduced peak power draw. The only caveat is because your aiming to time your heating to be within certain timing you'd probably want max possible power input during solar times and then drop down to minimal power at low solar times. My plan is to have a elevated temp increase during the day and allow a gentle taper at night during non-use times (11pm-5am). With a small bump up in the morning if I plan to use it.

**Reply by ChadH360** [reply]
I'm toying with the idea of disconnecting the airpump from the control panel, and then connecting it up to to an smart module that would allow me to switch it on/off manually or via Alexa (something like this - https://www.amazon.co.uk/gp/product/B089LWGNJ5 - which has low voltage on the switch side). I use these modules on a lot of my lighting circuits at home.
That way, I still have bubbles control, but my 'Bubbles' button on the pump unit then becomes a button to run the heater at half power instead of full power.
Might be an option for you, unless the unit throws an error.
I can't imagine it would though - surely it's just a relay that disconnects one of the heater elements and connects the air blower to the mains.

**Reply by admiralroflknife** [reply]
You are correct, the relay board is pretty basic and you can clearly see each one and could change the wiring to suit.

**Reply by visualapproach** [reply]
As I don't want anyone to be hurt while playing with this project, let me just remind you to use proper protection circuits. Water and mains voltage can be dangerous. I know you know that but I felt obligated to mention it.

**Reply by admiralroflknife** [reply]
Definitely! Safety is paramount with this for sure. The plug does have a GFCI built in luckily, and I have it in a smart plug that also has a GFCI function.... And the outlet it's plugged into has GFI... I might be overkilling it a little but yolo. 

I did end up fully disconnecting both wires off the one heater and wrapped the ends to ensure there's no voltage floating around. I encourage anyone else using this to be very careful as well.

**Reply by GSzabados** [reply]
Let me chime in as well. This is how the consumption of a Maldives looks like when the heater kicks in:

![image](https://user-images.githubusercontent.com/35445496/174476819-b1c62b71-234f-4c42-b05e-626a060d9c32.png)

It has a rush current at the beginning going up to 2150W and settles around 1850W. (middle of the graph) The circulation pump itself is around 50W.

I still haven't installed this project into the pump as I am waiting for it to run out of warranty first. But the pump is connected through a measuring plug and gives a good idea how much electricity is used during the summer.

I really like the idea of being able turning on only one of the heaters, as the solar inverter operates as a current generator, and produces equal current all three phases, so to cover the pump it needs to produce 6000W and that is not happening all day long. Direct consumption would be the best for this purpose.
